## Introduction

In [1]:
%%capture
%load_ext sql
%sql sqlite:///factbook.db

'Connected: None@factbook.db'

## Overview of the Data

The following will display a preview of the data so we can get a feel for it.

In [2]:
%%sql
SELECT *
  FROM sqlite_master
 WHERE type='table';

Done.


[('table', 'sqlite_sequence', 'sqlite_sequence', 3, 'CREATE TABLE sqlite_sequence(name,seq)'),
 ('table', 'facts', 'facts', 47, 'CREATE TABLE "facts" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "code" varchar(255) NOT NULL, "name" varchar(255) NOT NULL, "area" integer, "a ... (4 characters truncated) ... land" integer, "area_water" integer, "population" integer, "population_growth" float, "birth_rate" float, "death_rate" float, "migration_rate" float)')]

In [3]:
%%sql
SELECT *
    FROM facts
 LIMIT 5;

Done.


[(1, 'af', 'Afghanistan', 652230, 652230, 0, 32564342, 2.32, 38.57, 13.89, 1.51),
 (2, 'al', 'Albania', 28748, 27398, 1350, 3029278, 0.3, 12.92, 6.58, 3.3),
 (3, 'ag', 'Algeria', 2381741, 2381741, 0, 39542166, 1.84, 23.67, 4.31, 0.92),
 (4, 'an', 'Andorra', 468, 468, 0, 85580, 0.12, 8.13, 6.96, 0.0),
 (5, 'ao', 'Angola', 1246700, 1246700, 0, 19625353, 2.78, 38.78, 11.49, 0.46)]

Here are the descriptions for some of the columns:

* `name` — the name of the country.
* `area`— the country's total area (both land and water).
* `area_land` — the country's land area in square kilometers.
* `area_water` — the country's waterarea in square kilometers.
* `population` — the country's population.
* `population_growth`— the country's population growth as a percentage.
* `birth_rate` — the country's birth rate, or the number of births per year per 1,000 people.
* `death_rate` — the country's death rate, or the number of death per year per 1,000 people.

Let's start by calculating some summary statistics and look for any outlier countries.

In [5]:
%%sql

SELECT MIN(population) AS 'Minimum Population', MAX(population) AS 'Maximum Population', 
MIN(population_growth) AS 'Minimum Population Growth', MAX(population_growth) AS 'Maximum Population Growth'
    FROM facts;

Done.


[(0, 7256490011, 0.0, 4.02)]

We see a few interesting things in the summary statistics on the previous screen:

There's a country with a population of 0
There's a country with a population of 7256490011 (or more than 7.2 billion people)
Let's use subqueries to zoom in on just these countries without using the specific values.

## Exploring Outliers

In [6]:
%%sql

SELECT * 
    FROM facts
 WHERE population == (SELECT MIN(population) FROM facts);

Done.


[(250, 'ay', 'Antarctica', None, 280000, None, 0, None, None, None, None)]

In [7]:
%%sql

SELECT * 
    FROM facts
 WHERE population == (SELECT Max(population) FROM facts);

Done.


[(261, 'xx', 'World', None, None, None, 7256490011, 1.08, 18.6, 7.8, None)]

At first glance, one could guess that the 7.2 billion population was referring to the world. In regards to the populaton of 0, this is the continent of Anarctica.

In [10]:
%%sql

SELECT MIN(population) AS 'Minimum Population', 
MAX(population) AS 'Maximum Population', 
MIN(population_growth) AS 'Minimum Population Growth', 
MAX(population_growth) AS 'Maximum Population Growth'
    FROM facts
    WHERE name <> 'World';

Done.


[(0, 1367485388, 0.0, 4.02)]

In [14]:
%%sql
SELECT AVG(population) as Avg_pop, AVG(area) as Avg_area
    FROM facts
    WHERE name <> 'World';

Done.


[(32242666.56846473, 555093.546184739)]

In [17]:
%%sql

SELECT *
    FROM facts
    WHERE population > (SELECT AVG(population) FROM facts WHERE name <> 'World') AND 
    area < (SELECT AVG(area) FROM facts WHERE name <> 'World');


Done.


[(14, 'bg', 'Bangladesh', 148460, 130170, 18290, 168957745, 1.6, 21.14, 5.61, 0.46),
 (65, 'gm', 'Germany', 357022, 348672, 8350, 80854408, 0.17, 8.47, 11.42, 1.24),
 (80, 'iz', 'Iraq', 438317, 437367, 950, 37056169, 2.93, 31.45, 3.77, 1.62),
 (83, 'it', 'Italy', 301340, 294140, 7200, 61855120, 0.27, 8.74, 10.19, 4.1),
 (85, 'ja', 'Japan', 377915, 364485, 13430, 126919659, 0.16, 7.93, 9.51, 0.0),
 (91, 'ks', 'Korea, South', 99720, 96920, 2800, 49115196, 0.14, 8.19, 6.75, 0.0),
 (120, 'mo', 'Morocco', 446550, 446300, 250, 33322699, 1.0, 18.2, 4.81, 3.36),
 (138, 'rp', 'Philippines', 300000, 298170, 1830, 100998376, 1.61, 24.27, 6.11, 2.09),
 (139, 'pl', 'Poland', 312685, 304255, 8430, 38562189, 0.09, 9.74, 10.19, 0.46),
 (163, 'sp', 'Spain', 505370, 498980, 6390, 48146134, 0.89, 9.64, 9.04, 8.31),
 (173, 'th', 'Thailand', 513120, 510890, 2230, 67976405, 0.34, 11.19, 7.8, 0.0),
 (182, 'ug', 'Uganda', 241038, 197100, 43938, 37101745, 3.24, 43.79, 10.69, 0.74),
 (185, 'uk', 'United Kingdom', 243610, 241930, 1680, 64088222, 0.54, 12.17, 9.35, 2.54),
 (192, 'vm', 'Vietnam', 331210, 310070, 21140, 94348835, 0.97, 15.96, 5.93, 0.3)]

In [21]:
%%sql

SELECT name, MAX(population) as Highest_Population
    FROM facts
    WHERE name <> 'World';  

Done.


[('China', 1367485388)]

In [22]:
%%sql

SELECT name, MAX(population_growth) as Highest_Population_Growth
    FROM facts
    WHERE name <> 'World';

Done.


[('South Sudan', 4.02)]

In [31]:
%%sql

SELECT name, ROUND(CAST(area_water as Float) / CAST(area_land as Float), 2) as water_land_ratio
    FROM facts
    WHERE name <> 'World'
    ORDER BY water_land_ratio DESC;

Done.


[('British Indian Ocean Territory', 905.67),
 ('Virgin Islands', 4.52),
 ('Puerto Rico', 0.55),
 ('Bahamas, The', 0.39),
 ('Guinea-Bissau', 0.28),
 ('Malawi', 0.26),
 ('Netherlands', 0.23),
 ('Uganda', 0.22),
 ('Eritrea', 0.16),
 ('Liberia', 0.16),
 ('Bangladesh', 0.14),
 ('Gambia, The', 0.12),
 ('Taiwan', 0.12),
 ('Finland', 0.11),
 ('India', 0.11),
 ('Canada', 0.1),
 ('Colombia', 0.1),
 ('Sweden', 0.1),
 ('Brunei', 0.09),
 ('Guyana', 0.09),
 ('Nicaragua', 0.09),
 ('French Polynesia', 0.09),
 ('Burundi', 0.08),
 ('Iran', 0.08),
 ('Estonia', 0.07),
 ('Rwanda', 0.07),
 ('Tanzania', 0.07),
 ('United States', 0.07),
 ('Vietnam', 0.07),
 ('Norway', 0.06),
 ('Albania', 0.05),
 ('Armenia', 0.05),
 ('Azerbaijan', 0.05),
 ('Ghana', 0.05),
 ('Indonesia', 0.05),
 ('Suriname', 0.05),
 ('Tunisia', 0.05),
 ('Uzbekistan', 0.05),
 ('Burma', 0.04),
 ('Gabon', 0.04),
 ('Hungary', 0.04),
 ('Japan', 0.04),
 ('Kyrgyzstan', 0.04),
 ('Latvia', 0.04),
 ('Lithuania', 0.04),
 ('Romania', 0.04),
 ('Russia', 0.04),
 ('Togo', 0.04),
 ('Tonga', 0.04),
 ('Turkmenistan', 0.04),
 ('Ukraine', 0.04),
 ('West Bank', 0.04),
 ('Botswana', 0.03),
 ('Cambodia', 0.03),
 ('China', 0.03),
 ('Congo, Democratic Republic of the', 0.03),
 ('Iceland', 0.03),
 ('Korea, South', 0.03),
 ('Laos', 0.03),
 ('Moldova', 0.03),
 ('Montenegro', 0.03),
 ('Nepal', 0.03),
 ('Pakistan', 0.03),
 ('Poland', 0.03),
 ('Solomon Islands', 0.03),
 ('Switzerland', 0.03),
 ('Venezuela', 0.03),
 ('Hong Kong', 0.03),
 ('Argentina', 0.02),
 ('Austria', 0.02),
 ('Belarus', 0.02),
 ('Benin', 0.02),
 ('Brazil', 0.02),
 ('Bulgaria', 0.02),
 ('Chad', 0.02),
 ('Chile', 0.02),
 ('Czech Republic', 0.02),
 ('Denmark', 0.02),
 ('Ecuador', 0.02),
 ('El Salvador', 0.02),
 ('Germany', 0.02),
 ('Guatemala', 0.02),
 ('Ireland', 0.02),
 ('Israel', 0.02),
 ('Italy', 0.02),
 ('Kenya', 0.02),
 ('Lebanon', 0.02),
 ('Mali', 0.02),
 ('Mozambique', 0.02),
 ('Papua New Guinea', 0.02),
 ('Paraguay', 0.02),
 ('Saint Lucia', 0.02),
 ('Senegal', 0.02),
 ('Slovakia', 0.02),
 ('Somalia', 0.02),
 ('Sri Lanka', 0.02),
 ('Tajikistan', 0.02),
 ('Turkey', 0.02),
 ('New Caledonia', 0.02),
 ('Australia', 0.01),
 ('Belgium', 0.01),
 ('Belize', 0.01),
 ('Bolivia', 0.01),
 ('Cameroon', 0.01),
 ("Cote d'Ivoire", 0.01),
 ('Croatia', 0.01),
 ('Cuba', 0.01),
 ('Dominican Republic', 0.01),
 ('Egypt', 0.01),
 ('France', 0.01),
 ('Greece', 0.01),
 ('Haiti', 0.01),
 ('Jamaica', 0.01),
 ('Jordan', 0.01),
 ('Kazakhstan', 0.01),
 ('Macedonia', 0.01),
 ('Madagascar', 0.01),
 ('Mexico', 0.01),
 ('Mongolia', 0.01),
 ('Nigeria', 0.01),
 ('Panama', 0.01),
 ('Philippines', 0.01),
 ('Portugal', 0.01),
 ('Singapore', 0.01),
 ('Slovenia', 0.01),
 ('Spain', 0.01),
 ('Swaziland', 0.01),
 ('Syria', 0.01),
 ('United Kingdom', 0.01),
 ('Uruguay', 0.01),
 ('Zambia', 0.01),
 ('Zimbabwe', 0.01),
 ('Afghanistan', 0.0),
 ('Algeria', 0.0),
 ('Andorra', 0.0),
 ('Angola', 0.0),
 ('Antigua and Barbuda', 0.0),
 ('Bahrain', 0.0),
 ('Barbados', 0.0),
 ('Bhutan', 0.0),
 ('Bosnia and Herzegovina', 0.0),
 ('Burkina Faso', 0.0),
 ('Cabo Verde', 0.0),
 ('Central African Republic', 0.0),
 ('Comoros', 0.0),
 ('Congo, Republic of the', 0.0),
 ('Costa Rica', 0.0),
 ('Cyprus', 0.0),
 ('Djibouti', 0.0),
 ('Dominica', 0.0),
 ('Equatorial Guinea', 0.0),
 ('Fiji', 0.0),
 ('Georgia', 0.0),
 ('Grenada', 0.0),
 ('Guinea', 0.0),
 ('Honduras', 0.0),
 ('Iraq', 0.0),
 ('Kiribati', 0.0),
 ('Korea, North', 0.0),
 ('Kosovo', 0.0),
 ('Kuwait', 0.0),
 ('Lesotho', 0.0),
 ('Libya', 0.0),
 ('Liechtenstein', 0.0),
 ('Luxembourg', 0.0),
 ('Malaysia', 0.0),
 ('Maldives', 0.0),
 ('Malta', 0.0),
 ('Marshall Islands', 0.0),
 ('Mauritania', 0.0),
 ('Mauritius', 0.0),
 ('Micronesia, Federated States of', 0.0),
 ('Monaco', 0.0),
 ('Morocco', 0.0),
 ('Namibia', 0.0),
 ('Nauru', 0.0),
 ('Niger', 0.0),
 ('Oman', 0.0),
 ('Palau', 0.0),
 ('Peru', 0.0),
 ('Qatar', 0.0),
 ('Saint Kitts and Nevis', 0.0),
 ('Saint Vincent and the Grenadines', 0.0),
 ('Samoa', 0.0),
 ('San Marino', 0.0),
 ('Sao Tome and Principe', 0.0),


In [32]:
%%sql

SELECT name, ROUND(CAST(area_land as Float) / CAST(area_water as Float), 2) as 
land_to_water_ratio
    FROM facts
    WHERE name <> 'World'
    ORDER BY land_to_water_ratio DESC;

Done.


[('Bosnia and Herzegovina', 5118.7),
 ('Niger', 4222.33),
 ('Morocco', 1785.2),
 ('Guinea', 1755.12),
 ('Costa Rica', 1276.5),
 ('Djibouti', 1159.0),
 ('Korea, North', 926.22),
 ('Cyprus', 924.1),
 ('Namibia', 821.65),
 ('Burkina Faso', 684.5),
 ('Congo, Republic of the', 683.0),
 ('Sierra Leone', 596.83),
 ('Honduras', 559.45),
 ('Iraq', 460.39),
 ('Samoa', 282.1),
 ('Malaysia', 276.18),
 ('South Africa', 262.87),
 ('Peru', 245.21),
 ('Thailand', 229.1),
 ('Mauritius', 203.0),
 ('France', 189.81),
 ('Cameroon', 173.15),
 ('Egypt', 165.91),
 ('Slovenia', 165.17),
 ('Jordan', 164.45),
 ('Philippines', 162.93),
 ('Portugal', 147.53),
 ('Mongolia', 147.12),
 ('Uruguay', 145.85),
 ('Haiti', 145.05),
 ('United Kingdom', 144.01),
 ('Belize', 142.54),
 ('Dominican Republic', 138.06),
 ('Australia', 130.39),
 ('Belgium', 121.11),
 ('Syria', 118.47),
 ('Swaziland', 107.53),
 ('Kazakhstan', 107.13),
 ('Madagascar', 105.72),
 ('Cuba', 105.6),
 ('Greece', 99.73),
 ('Zimbabwe', 98.94),
 ('Mexico', 95.15),
 ('Macedonia', 90.83),
 ('Croatia', 90.28),
 ('Zambia', 80.63),
 ('Spain', 78.09),
 ("Cote d'Ivoire", 71.3),
 ('Bolivia', 70.9),
 ('Nigeria', 70.06),
 ('Panama', 68.83),
 ('Singapore', 68.7),
 ('Jamaica', 67.69),
 ('Sri Lanka', 65.95),
 ('El Salvador', 64.75),
 ('Denmark', 64.29),
 ('Argentina', 62.61),
 ('Guatemala', 61.94),
 ('Mali', 61.0),
 ('New Caledonia', 60.92),
 ('Somalia', 60.79),
 ('Saint Lucia', 60.6),
 ('Chile', 60.52),
 ('Mozambique', 60.49),
 ('Lebanon', 60.18),
 ('Austria', 57.82),
 ('Benin', 55.31),
 ('Turkey', 55.25),
 ('Tajikistan', 54.64),
 ('Brazil', 53.02),
 ('Slovakia', 51.73),
 ('Chad', 50.77),
 ('Kenya', 50.69),
 ('Ireland', 49.56),
 ('Czech Republic', 47.68),
 ('Israel', 46.2),
 ('Senegal', 45.93),
 ('Bulgaria', 45.39),
 ('Papua New Guinea', 45.38),
 ('Belarus', 43.17),
 ('Paraguay', 42.04),
 ('Germany', 41.76),
 ('Ecuador', 41.2),
 ('Italy', 40.85),
 ('Cambodia', 39.05),
 ('Laos', 38.47),
 ('Botswana', 37.78),
 ('Nepal', 37.43),
 ('Montenegro', 37.37),
 ('Iceland', 36.45),
 ('Poland', 36.09),
 ('Korea, South', 34.61),
 ('China', 34.47),
 ('Moldova', 34.26),
 ('Switzerland', 31.25),
 ('Solomon Islands', 30.75),
 ('Hong Kong', 30.66),
 ('Pakistan', 30.57),
 ('Venezuela', 29.4),
 ('Congo, Democratic Republic of the', 29.14),
 ('Burma', 28.33),
 ('Japan', 27.14),
 ('Romania', 27.05),
 ('Latvia', 26.6),
 ('Hungary', 26.2),
 ('Turkmenistan', 25.86),
 ('Gabon', 25.77),
 ('West Bank', 25.64),
 ('Lithuania', 23.92),
 ('Ukraine', 23.92),
 ('Tonga', 23.9),
 ('Kyrgyzstan', 23.53),
 ('Russia', 22.73),
 ('Togo', 22.66),
 ('Azerbaijan', 20.81),
 ('Ghana', 20.68),
 ('Albania', 20.29),
 ('Suriname', 19.95),
 ('Indonesia', 19.48),
 ('Uzbekistan', 19.34),
 ('Tunisia', 18.83),
 ('Armenia', 18.31),
 ('Norway', 15.59),
 ('Estonia', 14.93),
 ('Rwanda', 14.77),
 ('Vietnam', 14.67),
 ('Tanzania', 14.4),
 ('United States', 13.78),
 ('Iran', 13.14),
 ('Burundi', 11.94),
 ('Nicaragua', 11.56),
 ('French Polynesia', 11.26),
 ('Guyana', 10.86),
 ('Brunei', 10.53),
 ('Colombia', 10.37),
 ('Sweden', 10.27),
 ('Canada', 10.2),
 ('India', 9.47),
 ('Finland', 8.85),
 ('Taiwan', 8.67),
 ('Gambia, The', 8.58),
 ('Bangladesh', 7.12),
 ('Liberia', 6.4),
 ('Eritrea', 6.08),
 ('Uganda', 4.49),
 ('Netherlands', 4.43),
 ('Malawi', 3.86),
 ('Guinea-Bissau', 3.51),
 ('Bahamas, The', 2.59),
 ('Puerto Rico', 1.8),
 ('Virgin Islands', 0.22),
 ('British Indian Ocean Territory', 0.0),
 ('Afghanistan', None),
 ('Algeria', None),
 ('Andorra', None),
 ('Angola', None),
 ('Antigua and Barbuda', None),
 ('Bahrain', None),
 ('Barbados', None),
 ('Bhutan', None),
 ('Cabo Verde', None),
 ('Central African Republic', None),
 ('Comoros', None),
 ('Dominica', None),
 ('Equatorial Guinea', None),
 ('Ethiopia', None),
 ('Fiji', None),
 ('Georgia', None),
 ('Grenada', None),
 ('Kiribati', None),
 ('Kosovo', None),
 ('Kuwait', None),
 ('Lesotho', None),
 ('Libya', None),
 ('Liechtenstein', None),
 ('Luxembourg', None),
 ('Maldives', None),
 ('Malta', None),
 ('Marshall Islands